keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+
+### Goals
+
+Introduce a new parameter format based on the JSON-schema.
+
+### Non-Goals
+
+Extend the current format to support new features.
+
+### Proposal Overview
+
+One part of the proposal deals the client-side of the new format. This includes:
+ - Introduce new API version (`apiVersion:kudo.dev/v1beta2`) for the operator package `params.yaml`. KUDO should be able to read both old and new formats.
+- Support updating nested values from the command line e.g. `kudo update ... -p some.nested.key=value`
+- Support updating nested values using values file e.g. `kudo update ... -P values.yaml`
+- Add nested parameters via. the wizard `kudo package add parameter`
+- Parameter listing aka `kudo package list parameters`. We need a new output format here - maybe a tree-table?
+- Possible additional validation for our custom attributes e.g. parameter `trigger`
+
+The server-side of the proposal includes:
+- Updating `Operator` and `Instance` CRDs to use JSON-schema as the parameter format  
+- Providing a conversion-webhook to convert in-cluster running operators to the new format 
+   
+### Package Format
+
+JSON-schema is a [well-described](http://json-schema.org/understanding-json-schema/) format. However, here is a little example showing an existing `ZOOKEEPER_URI` and its new counterpart `Zookeeper.URI`:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: ZOOKEEPER_URI
+    displayName: "zookeeper cluster URI"
+    description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+    default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: true
+```
+
+```yaml
+type: object
+"$schema": "http://json-schema.org/draft-07/schema#"
+apiVersion: kudo.dev/v1beta2
+properties:
+  Zookeeper:
+    properties:
+      URI:
+        type: "string"
+        description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+        default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: ["URI"]
+```
+
+KUDO will support both package formats for the future. At some point the old format might become deprecated, but this will not be planned yet.
+
+Note: The apiVersions here (`v1beta1` and `v2beta2`) correspond to the CRD versions, but are separate definitions, even if they seem to be similar in certain aspects.
+
+### CRDs
+
+Normally, per Kubernetes conventions, CRDs would evolve over released versions without a breaking change from one version to the next, allowing clients to gradually convert to new CRD versions. This requires for a CRD to be convertible from the old version to new _and vice versa_ for all clients (new and old) to be able to read/write both old and new CRD versions. However, such a conversion in both directions will not be possible with this change, as JSON-schema supports a lot of configurations which will not be mappable from a flat list. 
```

Comment:
I wonder if this is right, I have to do some reading. I wonder if we should not have really new version of API/different API when we're doing such a breaking change. Is one way conversion a norm? I guess I need to do some reading here to be able to evaluate.

Reply:
I think this is fine, this is what the docs say for beta features: `The schema and/or semantics of objects may change in incompatible ways in a subsequent beta or stable release. `

Reply:
But at the same time I found this: `In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information.`

Reply:
After that btw. they propose that you need to change v1 and enhance it in a way so that it can contain all the information from v2. Can you imagine doing that?

Reply:
Enhancing the `v1beta` version so it can contain all the data from `v1beta2` was my first idea, but that would mean we'd have to restrict the usage of json-schema to a quite limited subset. So, realistically, I don't see that as a viable option.

The `all versions must be safely round-tripable` is true, but only for all *served* versions. The current idea is to deprecate `v1beta` immediatly and only serve `v1beta2`. In the code, we still need to have `v1beta1`, but only for migration purposes.

This ensures that only conversions from `v1beta` to `v1beta2` are done, never the other way around.

Reply:
Alright, I cannot really evaluate how good that option is and if it poses some problems, I would have to do a lot of reading. All the sources I found always talk about round-tripability as main feature for all fields https://groups.google.com/g/kubernetes-sig-api-machinery/c/07JKqCzQKsc?pli=1 🤷

Reply:
do you have some examples/reading about someone who also did it this way? or basically any conversion webhooks in the wild?

Reply:
For examples: The only real project I know that uses multiple CRD versions is cert-manager, and I looked at their code for some of the things. They do register a conversion webhook, but fully rely on the default schema conversion without any custom code. I haven't looked into their CRDs to see how closely they follow the full-roundtrip rule. I assume close, as they might serve multiple/all versions.
I don't know any project that does "one-way-conversion" or migration though.

And yeah. The solution described here is not optimal and breaks the Kubernetes API compatibility rules, but it shouldn't impose any technical problems.

The alternatives would be one of these:
1) Make it a fully breaking change - Ignore existing installations and just change the CRD. I'm very opposed to this solution.
2) Do it following the K8s API rules - This would involve multiple new CRD versions over multiple KUDO releases and not really change much - The current released clients still won't work with the final CRDs, but there is a time period in between where we support both old and new model where clients can switch. I argue that we don't have many clients yet, and are not that integrated in existing structures, so we can get away with skipping this migration period.